File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    void solve(TreeNode* root, int maxo, int mino, int &ans) {
        if (root == nullptr) {
            ans = max(ans, abs(maxo - mino));
            return;
        }
        ans = max(ans, abs(maxo - mino));
        solve(root->left, max(maxo, root->val), min(mino, root->val), ans);
        solve(root->right, max(maxo, root->val), min(mino, root->val), ans);
    }

    int maxAncestorDiff(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        if (root->left == nullptr && root->right == nullptr) {
            return root->val;
        }
        int mino = root->val;
        int maxo = root->val;
        int ans = 0;
        solve(root , maxo , mino , ans);
        return ans;
    }
};

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int result;
    
    public int maxAncestorDiff(TreeNode root) {
        result = 0;
        solve(root, 1000000, -1);
        return result;
    }
    
    public void solve(TreeNode root, int min, int max) {
        if (root == null) {
            result = Math.max(result, max - min);
            return;
        }
        solve(root.left, Math.min(root.val, min), Math.max(root.val, max));
        solve(root.right, Math.min(root.val, min), Math.max(root.val, max));
        result = Math.max(result, max - min);
    }
}